lib/Onyx.d.ts

```diff
@@ -0,0 +1,243 @@
+ * to note that if you have an array value property on an `Object` that the default behavior of lodash/merge is not to
+ * concatenate. See here: https://github.com/lodash/lodash/issues/2872
+ *
+ * Calls to `Onyx.merge()` are batched so that any calls performed in a single tick will stack in a queue and get
+ * applied in the order they were called. Note: `Onyx.set()` calls do not work this way so use caution when mixing
+ * `Onyx.merge()` and `Onyx.set()`.
+ *
+ * @example
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Joe']); // -> ['Joe']
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Jack']); // -> ['Joe', 'Jack']
+ * Onyx.merge(ONYXKEYS.POLICY, {id: 1}); // -> {id: 1}
+ * Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Workspace'}
+ *
+ * @param {String} key ONYXKEYS key
+ * @param {(Object|Array)} value Object or Array value to merge
+ * @returns {Promise}
+ */
+declare function merge<K extends KeyList>(key: K, value: Partial<Values[K]>): Promise<void>;
+
+/**
+ * Clear out all the data in the store
+ *
+ * Note that calling Onyx.clear() and then Onyx.set() on a key with a default
+ * key state may store an unexpected value in Storage.
+ *
+ * E.g.
+ * Onyx.clear();
+ * Onyx.set(ONYXKEYS.DEFAULT_KEY, 'default');
+ * Storage.getItem(ONYXKEYS.DEFAULT_KEY)
+ *     .then((storedValue) => console.log(storedValue));
+ * null is logged instead of the expected 'default'
+ *
+ * Onyx.set() might call Storage.setItem() before Onyx.clear() calls
+ * Storage.setItem(). Use Onyx.merge() instead if possible. Onyx.merge() calls
+ * Onyx.get(key) before calling Storage.setItem() via Onyx.set().
+ * Storage.setItem() from Onyx.clear() will have already finished and the merged
+ * value will be saved to storage after the default value.
+ *
+ * @param keysToPreserve is a list of ONYXKEYS that should not be cleared with the rest of the data
+ */
+declare function clear<K extends KeyList>(keysToPreserve?: K[]): Promise<void>;
+
+/**
+ * Merges a collection based on their keys
+ *
+ * @example
+ *
+ * Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, {
+ *     [`${ONYXKEYS.COLLECTION.REPORT}1`]: report1,
+ *     [`${ONYXKEYS.COLLECTION.REPORT}2`]: report2,
+ * });
+ *
+ * @param {String} collectionKey e.g. `ONYXKEYS.COLLECTION.REPORT`
+ * @param {Object} collection Object collection keyed by individual collection member keys and values
+ * @returns {Promise}
+ */
+declare function mergeCollection<K extends KeyList>(collectionKey: K, collection: Record<K, Values[K]>): Promise<void>;
+
+/**
+ * Insert API responses and lifecycle data into Onyx
+ *
+ * @param {Array} data An array of objects with shape {onyxMethod: oneOf('set', 'merge', 'mergeCollection'), key: string, value: *}
+ * @returns {Promise} resolves when all operations are complete
+ */
+declare function update<K extends KeyList>(data: UpdateOperation<K>[]): Promise<void>;
```

Comment:
In TypeScript, it might be ambiguous to infer array types where each item might have different type. That's why we have to treat each element individually so that TS knows to narrow the type for each element! Let me know if it works for you 😄 

```ts
declare type UpdateOperation<K extends KeyList> = {
    onyxMethod: 'set' | 'merge' | 'mergeCollection';
    key: K;
    value: Values[K];
};

declare type UpdateOperations<OK extends KeyList[]> = {
    [K in keyof OK]: UpdateOperation<OK[K]>;
};

declare function update<OK extends KeyList[]>(data: UpdateOperations<[...OK]>): Promise<void>;
```

Reply:
I used type parameter `OK` because I have `OnyxKey` type in my code, feel free to change names @fabioh8010

Reply:
@blazejkustra Brilliant! 🎉

Reply:
@blazejkustra @fabioh8010 

I tried both approaches, and both works unless I'm missing something. You can try changing the arguments passed to `update1` and `update2` defined in [this TS playground ](https://www.typescriptlang.org/play?#code/C4TwDgpgBA0hJQLxQOQGt4EYVQD6oxACYd914BmFAbgChRIoA1AQwBsBXCAZySgG9aUKIUwAuKN2AAnAJYA7AOZ1hhIhMHDhAMwD2uiVLlKANEK0AjFtInyOAWwsRp5gL4mo5whQkX9bCBZ5AG0AXVpXWnpwaABVMAATFmAIAHlIaWTZXXkAHhgoCAAPFPkE3jgQAD4+fhF4CRgPADd2LglWTh5gmFCoVzoGOMTktIysnO5cgBVK3mLS8th4MJrkOp6oBXqQXW0oWfhuUIl4pJT05wm8w5BuHtCagajtDnkAY2Bs+SgOEZTMDM5oUShAyhUVo8ABR-c48U7-MZXL6TIFHKoASgE5nek10AQAdGxdIoYYjuBi6JFXh8UT9YaMiPkQYsIdUyXDuAi4ZdMnT8lUwljNFBcfJuPiIESSRzRhSqVEGQCocF+IQJOQQFQWm0IBJgtp2NwIB4ZFxQu4BOqCFgUDquhqABYQNjElCuUKU2hKiBEFVqho2rV2qCtB1QA1Gk1QM0QC0eAMgDWiENh9qoZ2u3Tuz3UIA).

The original approach `declare function update<K extends KeyList>(data: UpdateOperation<K>[]): Promise<void>;` seems simpler IMO

let me know if I'm missing something here

Reply:
> The original approach declare function update<K extends KeyList>(data: UpdateOperation<K>[]): Promise<void>; seems simpler IMO

Yes it is simpler but it doesn't provide type safety when you update multiple keys, here is an example:
```ts
type Key = 'str' | 'num' | 'bool';
type Value = {
    str: string;
    num: number;
    bool: boolean;
};

declare type UpdateOperation<K extends Key> = {
    onyxMethod: 'set' | 'merge' | 'mergeCollection';
    key: K;
    value: Value[K];
};

declare function update<K extends Key>(data: Array<UpdateOperation<K>>): Promise<void>;

update([
    {
        onyxMethod: 'set',
        key: 'str',
        value: 'test',
    },
    {
        onyxMethod: 'merge',
        key: 'num',
        value: 'str', // There is no error (value: string | number | boolean)
    },
    {
        onyxMethod: 'merge',
        key: 'bool',
        value: 1, // There is no error (value: string | number | boolean)
    },
]);
```

Reply:
ah now I got it. thank you @blazejkustra!